File: 0001-0100/0077.Combinations.cpp
//30ms 99.65%


class Solution {
public:

    int nCr(int n, int k) {
        k = min(k, n - k);
        int ans = 1;

        for (int i = 1; i <= k; i++) {
            ans *= n - k + i;
            ans /= i;
        }

        return ans;
    }

	vector<vector<int>> combine(int n, int k) {
		vector<vector<int>> result;
		int i = 0;
        result.reserve(nCr(n, k));
		vector<int> p(k);
		while (i >= 0) {
			p[i]++;
			if (p[i] > n) --i;
			else if (i == k - 1) result.push_back(p);
			else {
			    ++i;
			    p[i] = p[i - 1];
			}
		}
		return result;
	}
};

File: 0001-0100/0077.Combinations.java
//0ms

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> combine(int n, int k) {
        return new AbstractList() {
            List<List<Integer>> result;

            private void init() {
                result = new ArrayList<>();
                List<Integer> temp = new ArrayList<>();
                solve(n, 1, k, result, temp);
            }

            @Override
            public int size() {
                if (result == null) {
                    init();
                }
                return result.size();
            }

            @Override
            public Object get(int index) {
                return result.get(index);
            }
        };
    }


    private void solve(int n, int current, int k, List<List<Integer>> result, List<Integer> temp) {
        if (k == 0) {
            result.add(new ArrayList<>(temp));
            return;
        }

        for (int i = current; i <= n; i++) {
            temp.add(i);
            solve(n, i + 1, k - 1, result, temp);
            temp.removeLast();
        }

    }
}
